File: 프로그래머스/2/42584. 주식가격/주식가격.cpp
#include <string>
#include <vector>
#include <stack>

std::vector<int> solution(std::vector<int> prices) {
    std::vector<int> answers(prices.size());
    std::stack<int> index_stack;
    
    for (int idx = 0; idx < prices.size(); ++idx) {
        while (!index_stack.empty() && prices[index_stack.top()] > prices[idx]) {
            int top_index = index_stack.top();
            answers[top_index] = idx - top_index;
            index_stack.pop();
        }
        index_stack.push(idx);
    }
    
    while (!index_stack.empty()) {
        int top_index = index_stack.top();
        answers[top_index] = prices.size() - top_index - 1;
        index_stack.pop();
    }
    
    return answers;
}

File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.ArrayDeque;

class Solution {
    public int[] solution(int[] prices) {
        
        int n = prices.length;
        int[] durations = new int[n];
        
        ArrayDeque<Integer> stack = new ArrayDeque<>(n);
        stack.push(0);
        
        // 각 시점별로 가격이 떨어지지 않은 기간 계산
        for (int i = 1; i < n; i++) {
            // 현재 가격보다 높은 가격들은 여기서 가격이 떨어짐
            while (!stack.isEmpty() && prices[i] < prices[stack.peek()]) {
                int prevTime = stack.pop();
                durations[prevTime] = i - prevTime;
            }
            stack.push(i);
        }
        
        // 마지막까지 가격이 떨어지지 않은 기간 처리
        while (!stack.isEmpty()) {
            int time = stack.pop();
            durations[time] = n - 1 - time;
        }
        
        return durations;
    }
}